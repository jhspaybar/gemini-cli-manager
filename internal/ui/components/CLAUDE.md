# UI Components Development Guide

This guide provides best practices and patterns for developing reusable UI components in the Gemini CLI Manager.

## Overview

The UI components package (`internal/ui/components`) contains reusable, testable UI elements that can be shared across the application. This approach promotes:

- **Code reuse** - Write once, use everywhere
- **Consistency** - Uniform UI behavior across the app
- **Testability** - Components can be tested in isolation
- **Maintainability** - Changes in one place affect all uses

## Component Design Principles

### 1. Self-Contained
Each component should be a complete, independent unit:
```go
type Component struct {
    // Internal state
    data      []Item
    selected  int
    
    // Configuration
    width     int
    height    int
    
    // Styling
    styles    ComponentStyles
}
```

### 2. Configurable
Components should expose configuration methods:
```go
// Good: Chainable configuration
tabBar := components.NewTabBar(tabs, width).
    SetStyles(activeStyle, inactiveStyle, borderColor).
    SetActiveIndex(0)

// Good: Individual setters for runtime changes
tabBar.SetWidth(newWidth)
tabBar.SetActiveByID("settings")
```

### 3. Theme-Aware
Always use theme colors, never hardcode:
```go
// ‚ùå BAD
style := lipgloss.NewStyle().Foreground(lipgloss.Color("#7D56F4"))

// ‚úÖ GOOD
style := lipgloss.NewStyle().Foreground(theme.Primary())
```

### 4. Flexible Rendering
Provide multiple rendering options:
```go
// Basic render
output := component.Render()

// Render with additional context
output := component.RenderWithContent(content, height)

// Render specific parts
header := component.RenderHeader()
body := component.RenderBody()
```

## Creating a New Component

### 1. Define the Structure
```go
// Package declaration with documentation
package components

// ComponentName represents a reusable UI element
type ComponentName struct {
    // Public fields (if any)
    
    // Private fields
    items    []Item
    selected int
    width    int
    styles   Styles
}

// Item represents a single item in the component
type Item struct {
    ID    string
    Label string
    Icon  string
}
```

### 2. Provide Constructor
```go
// NewComponentName creates a new component instance
func NewComponentName(items []Item, width int) *ComponentName {
    return &ComponentName{
        items: items,
        width: width,
        // Set sensible defaults
        selected: 0,
        styles:   DefaultStyles(),
    }
}
```

### 3. Add Configuration Methods
```go
// SetWidth updates the component width
func (c *ComponentName) SetWidth(width int) {
    c.width = width
}

// SetStyles allows style customization
func (c *ComponentName) SetStyles(styles Styles) {
    c.styles = styles
}

// SetSelected sets the selected item by index
func (c *ComponentName) SetSelected(index int) error {
    if index < 0 || index >= len(c.items) {
        return fmt.Errorf("index out of bounds: %d", index)
    }
    c.selected = index
    return nil
}
```

### 4. Implement Render Methods
```go
// Render produces the component's visual output
func (c *ComponentName) Render() string {
    var output strings.Builder
    
    // Build the component
    for i, item := range c.items {
        style := c.styles.Normal
        if i == c.selected {
            style = c.styles.Selected
        }
        
        line := fmt.Sprintf("%s %s", item.Icon, item.Label)
        output.WriteString(style.Render(line))
        output.WriteString("\n")
    }
    
    return output.String()
}
```

## Testing Components

### 1. Unit Tests
Test component logic in isolation:
```go
func TestComponentName_SetSelected(t *testing.T) {
    items := []Item{
        {ID: "1", Label: "First"},
        {ID: "2", Label: "Second"},
    }
    comp := NewComponentName(items, 80)
    
    // Test valid selection
    err := comp.SetSelected(1)
    assert.NoError(t, err)
    assert.Equal(t, 1, comp.GetSelected())
    
    // Test invalid selection
    err = comp.SetSelected(5)
    assert.Error(t, err)
}
```

### 2. Visual Tests
Create test programs to verify appearance:
```go
// test/adhoc/test_component.go
func main() {
    theme.SetTheme("github-dark")
    
    items := generateTestItems()
    comp := components.NewComponentName(items, 80)
    
    // Test different states
    fmt.Println("Normal state:")
    fmt.Println(comp.Render())
    
    fmt.Println("\nWith selection:")
    comp.SetSelected(2)
    fmt.Println(comp.Render())
}
```

### 3. Integration Tests
Test how components work together:
```go
func TestTabBarWithList(t *testing.T) {
    tabs := []Tab{{Title: "Items", ID: "items"}}
    tabBar := NewTabBar(tabs, 80)
    
    list := NewList(items, 80)
    
    // Render together
    output := tabBar.RenderWithContent(list.Render(), 20)
    
    // Verify output
    assert.Contains(t, output, "Items")
    assert.Contains(t, output, "First item")
}
```

## Component Patterns

### 1. State Management
Keep state internal but provide access methods:
```go
type List struct {
    items    []Item
    cursor   int
    selected map[int]bool
}

// Public getters
func (l *List) GetCursor() int { return l.cursor }
func (l *List) IsSelected(index int) bool { return l.selected[index] }

// State modifiers return new state (immutable style)
func (l *List) MoveCursor(delta int) *List {
    newList := *l
    newList.cursor = clamp(l.cursor+delta, 0, len(l.items)-1)
    return &newList
}
```

### 2. Event Handling
For interactive components, define clear interfaces:
```go
// Event types
type SelectEvent struct{ Index int }
type ChangeEvent struct{ OldIndex, NewIndex int }

// Handler interface
type EventHandler interface {
    OnSelect(SelectEvent)
    OnChange(ChangeEvent)
}

// Component with handlers
func (c *Component) SetHandler(h EventHandler) {
    c.handler = h
}
```

### 3. Composition
Build complex components from simpler ones:
```go
type Form struct {
    title  *Text
    fields []*Input
    submit *Button
}

func (f *Form) Render() string {
    parts := []string{
        f.title.Render(),
        "",  // Empty line
    }
    
    for _, field := range f.fields {
        parts = append(parts, field.Render())
    }
    
    parts = append(parts, "", f.submit.Render())
    
    return strings.Join(parts, "\n")
}
```

## Best Practices

### 1. Width Calculations
Always account for borders and padding:
```go
func (c *Component) Render() string {
    // Account for borders (2) and padding (4)
    contentWidth := c.width - 6
    
    // Ensure minimum width
    if contentWidth < 10 {
        contentWidth = 10
    }
    
    // Use MaxWidth to prevent overflow
    return lipgloss.NewStyle().
        Border(lipgloss.RoundedBorder()).
        Padding(1, 2).
        Width(c.width).
        MaxWidth(c.width).
        Render(content)
}
```

### 2. Error Handling
Fail gracefully with sensible defaults:
```go
func (c *Component) Render() string {
    if len(c.items) == 0 {
        return c.renderEmptyState()
    }
    
    if c.width < MinimumWidth {
        c.width = MinimumWidth
    }
    
    // Normal rendering...
}
```

### 3. Performance
Cache expensive computations:
```go
type Table struct {
    data         [][]string
    columnWidths []int
    dirty        bool
}

func (t *Table) calculateColumnWidths() {
    if !t.dirty {
        return  // Use cached values
    }
    
    // Expensive calculation...
    t.dirty = false
}
```

### 4. Documentation
Document public APIs thoroughly:
```go
// TabBar renders a horizontal bar of selectable tabs.
// 
// The active tab is visually distinguished and connects seamlessly
// with content rendered below. Tabs can have icons and IDs for
// programmatic selection.
//
// Example:
//
//	tabs := []Tab{
//	    {Title: "Home", Icon: "üè†", ID: "home"},
//	    {Title: "Settings", Icon: "‚öôÔ∏è", ID: "settings"},
//	}
//	tabBar := NewTabBar(tabs, 80)
//	tabBar.SetActiveByID("home")
//	output := tabBar.Render()
type TabBar struct {
    // ...
}
```

## Component Checklist

When creating a new component, ensure:

- [ ] **Constructor** with sensible defaults
- [ ] **Configuration methods** for runtime changes
- [ ] **Theme integration** using theme package
- [ ] **Width handling** with proper calculations
- [ ] **Error handling** for edge cases
- [ ] **Documentation** with examples
- [ ] **Unit tests** for logic
- [ ] **Visual tests** in test/adhoc
- [ ] **README entry** in components/README.md

## Current Components

### TabBar
- **Purpose**: Renders horizontal tabs with content area
- **Features**: Dynamic tabs, icon support, seamless borders
- **Test files**: test/adhoc/test_tabs*.go

### Card
- **Purpose**: Renders cards for extensions, profiles, and other items
- **Features**: Multiple states (normal, selected, focused, active), icon/subtitle support, metadata display, truncation, responsive width
- **Test files**: 
  - Unit tests: internal/ui/components/card_test.go
  - Visual test: Run `make card` in test/adhoc/
- **Usage example**:
```go
card := components.NewCard(60).
    SetTitle("Extension Name", "üß©").
    SetSubtitle("v1.2.0").
    SetDescription("Extension description").
    AddMetadata("MCP Servers", "2 servers", "‚ö°").
    SetSelected(isSelected)

output := card.Render()
```

### (Future components)
- **Modal** - Centered modal dialogs
- **FormField** - Consistent form inputs
- **StatusBar** - Three-section status bar
- **EmptyState** - "No items" displays
- **SearchBar** - Already exists, needs to be moved to components

## Contributing

When adding new components:

1. Check if similar component exists
2. Discuss design in issue/PR
3. Follow patterns in this guide
4. Add comprehensive tests
5. Update documentation
6. Consider accessibility (screen reader friendly output)

Remember: Components should be **reusable**, **testable**, and **maintainable**.