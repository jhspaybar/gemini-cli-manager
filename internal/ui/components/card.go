package components

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/lipgloss"
	"github.com/jhspaybar/gemini-cli-manager/internal/theme"
)

// Card represents a reusable card UI component
type Card struct {
	width       int
	title       string
	titleIcon   string
	subtitle    string
	description string
	metadata    []MetadataItem
	selected    bool
	focused     bool
	active      bool // For states like "active profile"

	// Styles
	normalStyle   lipgloss.Style
	selectedStyle lipgloss.Style
	focusedStyle  lipgloss.Style
	activeStyle   lipgloss.Style
}

// MetadataItem represents a key-value pair in the card
type MetadataItem struct {
	Key   string
	Value string
	Icon  string // Optional icon
}

// NewCard creates a new card component
func NewCard(width int) *Card {
	// Default styles using theme
	borderColor := theme.Border()
	accentColor := theme.Primary()
	successColor := theme.Success()

	return &Card{
		width:    width,
		metadata: []MetadataItem{},

		normalStyle: lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(borderColor).
			Padding(1, 2).
			MaxWidth(width),

		selectedStyle: lipgloss.NewStyle().
			Border(lipgloss.ThickBorder()).
			BorderForeground(accentColor).
			Padding(1, 2).
			MaxWidth(width),

		focusedStyle: lipgloss.NewStyle().
			Border(lipgloss.DoubleBorder()).
			BorderForeground(accentColor).
			Padding(1, 2).
			MaxWidth(width),

		activeStyle: lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(successColor).
			Padding(1, 2).
			MaxWidth(width),
	}
}

// SetTitle sets the card title with optional icon
func (c *Card) SetTitle(title, icon string) *Card {
	c.title = title
	c.titleIcon = icon
	return c
}

// SetSubtitle sets a subtitle (e.g., version number)
func (c *Card) SetSubtitle(subtitle string) *Card {
	c.subtitle = subtitle
	return c
}

// SetDescription sets the card description
func (c *Card) SetDescription(description string) *Card {
	c.description = description
	return c
}

// AddMetadata adds a key-value metadata item with optional icon
func (c *Card) AddMetadata(key, value, icon string) *Card {
	c.metadata = append(c.metadata, MetadataItem{Key: key, Value: value, Icon: icon})
	return c
}

// SetSelected sets the selected state
func (c *Card) SetSelected(selected bool) *Card {
	c.selected = selected
	return c
}

// SetFocused sets the focused state
func (c *Card) SetFocused(focused bool) *Card {
	c.focused = focused
	return c
}

// SetActive sets the active state (e.g., for active profile)
func (c *Card) SetActive(active bool) *Card {
	c.active = active
	return c
}

// SetStyles allows custom style configuration
func (c *Card) SetStyles(normal, selected, focused, active lipgloss.Style) *Card {
	c.normalStyle = normal
	c.selectedStyle = selected
	c.focusedStyle = focused
	c.activeStyle = active
	return c
}

// SetWidth updates the card width
func (c *Card) SetWidth(width int) *Card {
	c.width = width
	// Update all styles with new width using MaxWidth
	c.normalStyle = c.normalStyle.MaxWidth(width)
	c.selectedStyle = c.selectedStyle.MaxWidth(width)
	c.focusedStyle = c.focusedStyle.MaxWidth(width)
	c.activeStyle = c.activeStyle.MaxWidth(width)
	return c
}

// Render produces the card's visual output
func (c *Card) Render() string {
	// Select appropriate style
	style := c.normalStyle
	if c.active {
		style = c.activeStyle
	} else if c.focused {
		style = c.focusedStyle
	} else if c.selected {
		style = c.selectedStyle
	}

	// Content will be automatically constrained by MaxWidth on the style

	// Build content
	var content []string
	
	// Create a style for content that respects the card's MaxWidth
	// We need to account for the padding (2*2=4) when setting content width
	contentStyle := lipgloss.NewStyle().Width(c.width - 4)

	// Title line with optional subtitle
	titleParts := []string{}
	
	// Add active indicator if active
	if c.active {
		titleParts = append(titleParts, "●")
	}
	
	// Add icon if present
	if c.titleIcon != "" {
		titleParts = append(titleParts, c.titleIcon)
	}
	
	// Add title
	titleStyle := lipgloss.NewStyle().Bold(true).Foreground(theme.TextPrimary())
	if c.selected || c.focused {
		titleStyle = titleStyle.Foreground(theme.Primary())
	} else if c.active {
		titleStyle = titleStyle.Foreground(theme.Success())
	}
	
	// Handle subtitle (like version)
	if c.subtitle != "" {
		subtitleStyle := lipgloss.NewStyle().Foreground(theme.TextMuted())
		subtitleText := subtitleStyle.Render(c.subtitle)
		
		// Build the title parts
		prefix := strings.Join(titleParts, " ")
		if prefix != "" {
			prefix += " "
		}
		
		// Create a title line with subtitle
		titleText := titleStyle.Render(c.title)
		fullTitle := prefix + titleText + "  " + subtitleText
		content = append(content, contentStyle.Render(fullTitle))
	} else {
		// No subtitle, just render title
		titleText := titleStyle.Render(c.title)
		titleParts = append(titleParts, titleText)
		titleLine := strings.Join(titleParts, " ")
		content = append(content, contentStyle.Render(titleLine))
	}

	// Description (if any)
	if c.description != "" {
		descStyle := lipgloss.NewStyle().Foreground(theme.TextSecondary()).Width(c.width - 4)
		content = append(content, descStyle.Render(c.description))
	}

	// Metadata section
	if len(c.metadata) > 0 {
		for _, meta := range c.metadata {
			var metaLine string
			if meta.Icon != "" {
				metaLine = fmt.Sprintf("%s %s", meta.Icon, meta.Value)
			} else {
				metaLine = fmt.Sprintf("%s: %s", meta.Key, meta.Value)
			}
			content = append(content, contentStyle.Copy().Foreground(theme.Primary()).Render(metaLine))
		}
	}

	// Join content with newlines
	return style.Render(strings.Join(content, "\n"))
}

// RenderCompact renders a more compact version of the card
func (c *Card) RenderCompact() string {
	// Use compact styles (less padding, use MaxWidth)
	compactStyle := c.normalStyle.Copy().Padding(0, 1).MaxWidth(c.width)
	if c.active {
		compactStyle = c.activeStyle.Copy().Padding(0, 1).MaxWidth(c.width)
	} else if c.focused {
		compactStyle = c.focusedStyle.Copy().Padding(0, 1).MaxWidth(c.width)
	} else if c.selected {
		compactStyle = c.selectedStyle.Copy().Padding(0, 1).MaxWidth(c.width)
	}

	// Build compact title line
	parts := []string{}
	if c.active {
		parts = append(parts, "●")
	}
	if c.titleIcon != "" {
		parts = append(parts, c.titleIcon)
	}
	parts = append(parts, c.title)
	
	titleLine := strings.Join(parts, " ")
	
	return compactStyle.Render(titleLine)
}

// Helper function to pluralize
func pluralize(count int) string {
	if count == 1 {
		return ""
	}
	return "s"
}