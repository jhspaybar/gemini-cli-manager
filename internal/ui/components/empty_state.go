package components

import (
	"github.com/charmbracelet/lipgloss"
	"github.com/jhspaybar/gemini-cli-manager/internal/theme"
)

// EmptyState represents a component for displaying empty/no-data states
type EmptyState struct {
	width       int
	icon        string
	title       string
	description string
	action      string
	centered    bool
	style       lipgloss.Style
	iconStyle   lipgloss.Style
	titleStyle  lipgloss.Style
	descStyle   lipgloss.Style
	actionStyle lipgloss.Style
}

// NewEmptyState creates a new empty state component
func NewEmptyState(width int) *EmptyState {
	return &EmptyState{
		width:    width,
		centered: true,
		style: lipgloss.NewStyle().
			Border(lipgloss.RoundedBorder()).
			BorderForeground(theme.Border()).
			Padding(2, 4).
			Align(lipgloss.Center),
		iconStyle: lipgloss.NewStyle().
			Foreground(theme.TextSecondary()),
		titleStyle: lipgloss.NewStyle().
			Foreground(theme.TextPrimary()).
			Bold(true),
		descStyle: lipgloss.NewStyle().
			Foreground(theme.TextSecondary()),
		actionStyle: lipgloss.NewStyle().
			Foreground(theme.Primary()).
			Italic(true),
	}
}

// SetIcon sets the emoji/icon for the empty state
func (es *EmptyState) SetIcon(icon string) *EmptyState {
	es.icon = icon
	return es
}

// SetTitle sets the main message
func (es *EmptyState) SetTitle(title string) *EmptyState {
	es.title = title
	return es
}

// SetDescription sets the descriptive text
func (es *EmptyState) SetDescription(description string) *EmptyState {
	es.description = description
	return es
}

// SetAction sets the action hint (e.g., "Press 'n' to create")
func (es *EmptyState) SetAction(action string) *EmptyState {
	es.action = action
	return es
}

// SetCentered controls whether the empty state is centered
func (es *EmptyState) SetCentered(centered bool) *EmptyState {
	es.centered = centered
	return es
}

// SetWidth updates the component width
func (es *EmptyState) SetWidth(width int) *EmptyState {
	es.width = width
	return es
}

// SetStyles allows customization of all styles
func (es *EmptyState) SetStyles(boxStyle, iconStyle, titleStyle, descStyle, actionStyle lipgloss.Style) *EmptyState {
	es.style = boxStyle
	es.iconStyle = iconStyle
	es.titleStyle = titleStyle
	es.descStyle = descStyle
	es.actionStyle = actionStyle
	return es
}

// Render renders the empty state
func (es *EmptyState) Render() string {
	// Build content parts
	var parts []string
	
	if es.icon != "" {
		parts = append(parts, es.iconStyle.Render(es.icon))
		parts = append(parts, "") // Empty line after icon
	}
	
	if es.title != "" {
		parts = append(parts, es.titleStyle.Render(es.title))
	}
	
	if es.description != "" {
		parts = append(parts, es.descStyle.Render(es.description))
	}
	
	if es.action != "" {
		parts = append(parts, es.actionStyle.Render(es.action))
	}
	
	// Join parts vertically with center alignment
	content := lipgloss.JoinVertical(lipgloss.Center, parts...)
	
	// Calculate box width - constrain to reasonable size
	boxWidth := es.width
	if boxWidth > 60 {
		boxWidth = 60 // Max width for readability
	}
	if boxWidth < 30 {
		boxWidth = 30 // Min width
	}
	
	// Apply box styling
	// Use MaxWidth to constrain content, let lipgloss handle total width with borders
	box := es.style.
		MaxWidth(boxWidth).
		Render(content)
	
	// Center the box if requested
	if es.centered {
		// Get actual rendered width including borders
		actualWidth := lipgloss.Width(box)
		if es.width > actualWidth {
			return lipgloss.Place(es.width, lipgloss.Height(box), lipgloss.Center, lipgloss.Center, box)
		}
	}
	
	return box
}

// Preset empty states for common scenarios

// NoItemsFound creates an empty state for search with no results
func (es *EmptyState) NoItemsFound() *EmptyState {
	return es.
		SetIcon("🔍").
		SetTitle("No items found").
		SetDescription("Try adjusting your search or filters")
}

// NoData creates a generic no data empty state
func (es *EmptyState) NoData(itemType string) *EmptyState {
	return es.
		SetIcon("📭").
		SetTitle("No " + itemType).
		SetDescription("Nothing to display yet")
}

// ComingSoon creates an empty state for features not yet implemented
func (es *EmptyState) ComingSoon() *EmptyState {
	return es.
		SetIcon("🚧").
		SetTitle("Coming Soon").
		SetDescription("This feature is under construction")
}