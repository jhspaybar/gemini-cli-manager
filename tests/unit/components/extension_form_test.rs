#[cfg(test)]
mod tests {
    use crate::test_utils::*;
    use crossterm::event::{KeyCode, KeyEvent, KeyEventKind};
    use gemini_cli_manager::components::Component;
    use gemini_cli_manager::components::extension_form::{ExtensionForm, FormField};
    use insta::assert_snapshot;

    fn create_key_event(code: KeyCode) -> gemini_cli_manager::tui::Event {
        use crossterm::event::KeyModifiers;
        gemini_cli_manager::tui::Event::Key(KeyEvent {
            code,
            modifiers: KeyModifiers::NONE,
            kind: KeyEventKind::Press,
            state: crossterm::event::KeyEventState::NONE,
        })
    }

    fn create_test_form() -> ExtensionForm {
        let storage = create_test_storage();
        ExtensionForm::new(storage)
    }

    fn create_edit_form(_id: &str) -> ExtensionForm {
        let storage = create_test_storage();

        // Add test extension to storage
        let ext = ExtensionBuilder::new("Test Extension")
            .with_version("1.0.0")
            .with_description("Test description")
            .with_tags(vec!["test", "example"])
            .build();
        storage.save_extension(&ext).unwrap();

        ExtensionForm::with_extension(storage, &ext)
    }

    #[test]
    fn test_form_initial_state() {
        let form = create_test_form();

        assert!(!form.is_edit_mode());
        assert_eq!(form.current_field(), &FormField::Name);
        assert!(form.name_input().value().is_empty());
        assert_eq!(form.version_input().value(), "1.0.0"); // Default version
        assert!(form.description_input().value().is_empty());
        assert!(form.tags_input().value().is_empty());
        assert_eq!(form.context_file_name_input().value(), "GEMINI.md"); // Default context file
    }

    #[test]
    fn test_form_rendering() {
        let mut form = create_test_form();
        let mut terminal = setup_test_terminal(80, 30).unwrap();

        terminal
            .draw(|f| {
                form.draw(f, f.area()).unwrap();
            })
            .unwrap();

        // Verify form elements are rendered
        assert_buffer_contains(&terminal, "Create New Extension");
        assert_buffer_contains(&terminal, "Name");
        assert_buffer_contains(&terminal, "Version");
        assert_buffer_contains(&terminal, "Description");
        assert_buffer_contains(&terminal, "Tags");
    }

    #[test]
    fn test_edit_mode() {
        let form = create_edit_form("test-extension");

        assert!(form.is_edit_mode());
        assert_eq!(form.name_input().value(), "Test Extension");
        assert_eq!(form.version_input().value(), "1.0.0");
        assert_eq!(form.description_input().value(), "Test description");
        assert_eq!(form.tags_input().value(), "test, example");
    }

    #[test]
    fn test_field_navigation() {
        let mut form = create_test_form();

        // Start at Name
        assert_eq!(form.current_field(), &FormField::Name);

        // Tab to Version
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        assert_eq!(form.current_field(), &FormField::Version);

        // Tab to Description
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        assert_eq!(form.current_field(), &FormField::Description);

        // Shift+Tab back to Version
        form.handle_events(Some(create_key_event(KeyCode::BackTab)))
            .unwrap();
        assert_eq!(form.current_field(), &FormField::Version);
    }

    #[test]
    fn test_text_input() {
        let mut form = create_test_form();

        // Type name
        for ch in "My Extension".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }
        assert_eq!(form.name_input().value(), "My Extension");

        // Tab to version
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        // Clear default version
        for _ in 0.."1.0.0".len() {
            form.handle_events(Some(create_key_event(KeyCode::Backspace)))
                .unwrap();
        }
        // Type new version
        for ch in "2.0.0".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }
        assert_eq!(form.version_input().value(), "2.0.0");
    }

    #[test]
    fn test_validation_empty_name() {
        let mut form = create_test_form();

        // Try to save with empty name
        form.handle_events(Some(create_key_event(KeyCode::Char('s').into())))
            .unwrap();

        // Should show error
        let mut terminal = setup_test_terminal(80, 30).unwrap();
        terminal
            .draw(|f| {
                form.draw(f, f.area()).unwrap();
            })
            .unwrap();

        // Note: We can't directly test validation errors without app state access
        // The component might handle this differently
    }

    #[test]
    fn test_form_cancel() {
        let mut form = create_test_form();

        // Type some data
        for ch in "Test".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        // Press Escape to cancel
        let result = form
            .handle_events(Some(create_key_event(KeyCode::Esc)))
            .unwrap();

        // Should return a navigation action
        assert!(result.is_some());
    }

    #[test]
    fn test_multiline_input_description() {
        let mut form = create_test_form();

        // Navigate to description
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();

        // Type multiline text
        for ch in "Line 1".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        // Note: Multiline input might be handled differently by tui-input
        assert!(form.description_input().value().contains("Line 1"));
    }

    #[test]
    fn test_form_save() {
        let mut form = create_test_form();

        // Fill out form
        for ch in "New Extension".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        for ch in "2.0.0".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        // Save with Ctrl+S
        let key_event = KeyEvent {
            code: KeyCode::Char('s'),
            modifiers: crossterm::event::KeyModifiers::CONTROL,
            kind: KeyEventKind::Press,
            state: crossterm::event::KeyEventState::NONE,
        };
        let result = form
            .handle_events(Some(gemini_cli_manager::tui::Event::Key(key_event)))
            .unwrap();

        // Should return save action
        assert!(result.is_some());
    }

    #[test]
    fn test_context_file_fields() {
        let mut form = create_test_form();

        // Navigate to context file name field
        // Name -> Version -> Description -> Tags -> ContextFileName
        for _ in 0..4 {
            form.handle_events(Some(create_key_event(KeyCode::Tab)))
                .unwrap();
        }
        assert_eq!(form.current_field(), &FormField::ContextFileName);

        // Clear default and type new context file name
        // First clear the default value
        for _ in 0.."GEMINI.md".len() {
            form.handle_events(Some(create_key_event(KeyCode::Backspace)))
                .unwrap();
        }
        // Then type new name
        for ch in "README.md".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }
        assert_eq!(form.context_file_name_input().value(), "README.md");

        // Tab to context content
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        assert_eq!(form.current_field(), &FormField::ContextContent);

        // Type content
        for ch in "# Extension Documentation".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }
        assert!(
            form.context_content_input()
                .value()
                .contains("Extension Documentation")
        );
    }

    #[test]
    fn test_form_responsive() {
        let mut form = create_test_form();

        // Test various terminal sizes
        let sizes = vec![(60, 20), (80, 30), (120, 40), (40, 15)];

        for (width, height) in sizes {
            let mut terminal = setup_test_terminal(width, height).unwrap();

            // Should render without panic
            let result = terminal.draw(|f| {
                form.draw(f, f.area()).unwrap();
            });

            assert!(result.is_ok(), "Failed to render at {}x{}", width, height);
        }
    }

    #[test]
    fn test_snapshot_empty_form() {
        let mut form = create_test_form();
        let output = render_to_string(80, 30, |f| {
            form.draw(f, f.area()).unwrap();
        });

        assert_snapshot!(output.unwrap());
    }

    #[test]
    fn test_snapshot_filled_form() {
        let mut form = create_test_form();

        // Fill out all fields
        for ch in "My Awesome Extension".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        for ch in "3.2.1".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        for ch in "This is a comprehensive extension for productivity".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        // Skip to tags
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        form.handle_events(Some(create_key_event(KeyCode::Tab)))
            .unwrap();
        for ch in "productivity, ai, development".chars() {
            form.handle_events(Some(create_key_event(KeyCode::Char(ch))))
                .unwrap();
        }

        let output = render_to_string(80, 30, |f| {
            form.draw(f, f.area()).unwrap();
        });

        assert_snapshot!(output.unwrap());
    }
}
