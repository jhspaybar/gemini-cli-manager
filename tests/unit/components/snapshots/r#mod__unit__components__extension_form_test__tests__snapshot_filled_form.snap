---
source: tests/unit/components/extension_form_test.rs
expression: output.unwrap()
---
╭──────────────────────────── Create New Extension ────────────────────────────╮
│                                                                              │
│ ┌Name───────────────────────────────────────────────┐┌Version──────────────┐ │
│ │My Awesome Extension                               ││1.0.03.2.1           │ │
│ └───────────────────────────────────────────────────┘└─────────────────────┘ │
│ ┌Description────────────────────────────────────────┐┌Tags─────────────────┐ │
│ │This is a comprehensive extension for productivity ││                     │ │
│ └───────────────────────────────────────────────────┘└─────────────────────┘ │
│ ┌Context File────────────────────────┐┌MCP Servers (↑/↓ to navigate, n: new┐ │
│ │GEMINI.md                           ││  No MCP servers configured         │ │
│ └────────────────────────────────────┘│                                    │ │
│ ┌Context Content (Markdown)──────────┐│                                    │ │
│ │productivity, ai, development       ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ │                                    ││                                    │ │
│ └────────────────────────────────────┘└────────────────────────────────────┘ │
│   Tab: Next field | ↑/↓: Scroll | Type to edit | Ctrl+S: Save | Esc: Cancel  │
│                                                                              │
│                                                                              │
╰──────────────────────────────────────────────────────────────────────────────╯
